src/crimson/os/seastore/segment_cleaner.h

```diff
@@ -0,0 +1,169 @@
+    size_t target_journal_segments = 0;
+    size_t max_journal_segments = 0;
+
+    static config_t default_from_segment_manager(
+      SegmentManager &manager) {
+      return config_t{
+	manager.get_num_segments(),
+	static_cast<size_t>(manager.get_segment_size()),
+	2,
+	4};
+    }
+  };
+
+  /// Callback interface for querying and operating on segments
+  class ExtentCallbackInterface {
+  public:
+    /**
+     * get_next_dirty_extent
+     *
+     * returns all extents with dirty_from < bound
+     */
+    using get_next_dirty_extents_ertr = crimson::errorator<>;
+    using get_next_dirty_extents_ret = get_next_dirty_extents_ertr::future<
+      std::list<CachedExtentRef>>;
+    virtual get_next_dirty_extents_ret get_next_dirty_extents(
+      journal_seq_t bound ///< [in] return extents with dirty_from < bound
+    ) = 0;
+
+    /**
+     * rewrite_extent
+     *
+     * Updates t with operations moving the passed extents to a new
+     * segment.  Segments in extents_to_relocate may be invalid,
+     * implementation must correctly handle finding the current
+     * instance of any that are still live and ignore any that are
+     * no longer live.
+     */
+    using rewrite_extent_ertr = crimson::errorator<
+      crimson::ct_error::input_output_error>;
+    using rewrite_extent_ret = rewrite_extent_ertr::future<>;
+    virtual rewrite_extent_ret rewrite_extent(
+      Transaction &t,
+      CachedExtentRef extent) = 0;
+  };
+
+private:
+  segment_id_t next = 0;
+  config_t config;
+
+  journal_seq_t journal_tail_target;
+  journal_seq_t journal_tail_committed;
+  journal_seq_t journal_head;
+
+  ExtentCallbackInterface *ecb = nullptr;
+
+public:
+  SegmentCleaner(config_t config)
+    : config(config) {}
+
+  get_segment_ret get_segment() final;
+
+  // hack for testing until we get real space handling
+  void set_next(segment_id_t _next) {
+    next = _next;
+  }
+  segment_id_t get_next() const {
+    return next;
+  }
+
+
+  void put_segment(segment_id_t segment) final;
+
+  journal_seq_t get_journal_tail_target() const final {
+    return journal_tail_target;
+  }
+
+  void update_journal_tail_committed(journal_seq_t committed) final;
+
+  void update_journal_tail_target(journal_seq_t target);
+
+  void init_journal_tail(journal_seq_t tail) {
+    journal_tail_target = journal_tail_committed = tail;
+  }
+
+  void set_journal_head(journal_seq_t head) {
+    assert(journal_head == journal_seq_t() || head >= journal_head);
+    journal_head = head;
+  }
+
+  void set_extent_callback(ExtentCallbackInterface *cb) {
+    ecb = cb;
+  }
+
+  /**
+   * do_immediate_work
+   *
+   * Should be invoked prior to submission of any transaction,
+   * will piggy-back work required to maintain deferred work
+   * constraints.
+   */
+  using do_immediate_work_ertr = crimson::errorator<
+    crimson::ct_error::input_output_error>;
+  using do_immediate_work_ret = do_immediate_work_ertr::future<>;
+  do_immediate_work_ret do_immediate_work(
```

Comment:
nit, i'd suggest s/do_immediate_work/do_now/ . and s/do_deferred_work/do_later/ for shorter names.

Reply:
No, I prefer these names.